clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/VeniceAdaptiveIngestionThrottler.java

```diff
@@ -0,0 +1,125 @@
+package com.linkedin.davinci.kafka.consumer;
+
+import static com.linkedin.davinci.kafka.consumer.ConsumerPoolType.NON_CURRENT_VERSION_AA_WC_LEADER_POOL;
+import static com.linkedin.davinci.kafka.consumer.ConsumerPoolType.NON_CURRENT_VERSION_NON_AA_WC_LEADER_POOL;
+
+import com.linkedin.alpini.base.concurrency.Executors;
+import com.linkedin.davinci.config.VeniceServerConfig;
+import com.linkedin.venice.throttle.EventThrottler;
+import com.linkedin.venice.utils.DaemonThreadFactory;
+import java.text.DecimalFormat;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.function.BooleanSupplier;
+import java.util.function.Supplier;
+
+
+public class VeniceAdaptiveIngestionThrottler extends IngestionThrottler {
```

Comment:
Instead of extending IngestionThrottler which contains many throttler for different pool, we can consider wrapping around EventThrottler. If EventThrottler does not support updating limit, we can pre-build a bunch of them with different levels of throttling numbers and adjust the reference internally. In this case, in the IngestionThrottler, we can substitute any single pool EventThrottler with the new AdaptiveThrottler if config is enabled.